### PR TITLE
[shelly] Improved Motion Support, Support CoIoT Unicast, fixes

### PR DESCRIPTION
**Pull request by markus7017**
The PR adds some improvements
- New channels device#sensorSleepTime, sensor#motionAction for Shelly Motion, sensor#motionTimestamp fixed on CoAP updazes
- Support for CoIoT Unicast (the Motion doesn't support Multicast anymore), works also for other devices
- H&T device#charger fixed
- Version strong was not updated on RESTARTED in call cases (timing)
- some other optimizations

### CODE REVIEW

**Comment by markus7017** [comment]
@fwolter Allterco updated their firmware, which brings an important feature - helps the users and reduces my support. I want to include this in this PR.

Nevertheless, could you please have a first look to the Code and give feedback on code structure. The is a bigger one and adds a complete new component. I would work on your feedback before starting the review. So far I set the WIP label.

**Comment by fwolter** [comment]
If these are features which are independent from each other, it would be better to split these and file another PR for the updated firmware.

**Comment by markus7017** [comment]
@fwolter It's one new component Shelly Manager, which has SOME dependencies/changes in the base code, but the new Manager component depends on the changes in the base code. It's not "10 independent features in one PR"

**Comment by fwolter** [comment]
Ok. As the backlog is quite extensive, I'm afraid I can't review this in the short-term.

**Comment by markus7017** [comment]
@fwolter would it help if I split it 
- Part I: Base changes and supporting stuff (html+images, updated readme) - yes, it goes into the bundle, but doesn't get visible to the user
- Part II: Actual Shelly Manager component

I estimate Part I is about 1/3rd of the total

**Comment by fwolter** [comment]
It makes sense if the two parts are useable from each other independently. If the changes in the first part are useless without part II, it doesn't make sense to split it.

**Comment by markus7017** [comment]
@fwolter I reduced the scope of this PR to include regular changes, but not the Manager
I kindly request to do the review on short-term so I get out of the sitioation that the PR doesn't matches the current development and I need to keep this branch up-to-date

**Comment by markus7017** [comment]
@fwolter does that work for you?

File: bundles/org.openhab.binding.shelly/README.md [hunk 1]
```diff
@@ -793,6 +802,11 @@ You can define 2 items (1 Switch, 1 Number) mapping to the same channel, see exa
 
 ### Shelly Motion (thing-type: shellymotion)
 
+Important: The Shelly Motion does only support CoIoT Unicast, which means you need to set the CoIoT peer address.
+
+- Use device WebUI, open COIOT settings, make sure CoIoT is enabled and enter the openHAB IP address or
+- Use [Shelly Manager](doc/ShellyManager.md, select Action 'Set CoIoT peer' and the Manager will sets the openHAB IP address as peer address
```
Comment (fwolter) [inline bundles/org.openhab.binding.shelly/README.md:808]:
```suggestion
- Use [Shelly Manager](doc/ShellyManager.md), select Action 'Set CoIoT peer' and the Manager will sets the openHAB IP address as peer address
```

File: bundles/org.openhab.binding.shelly/README.md [hunk 2]
```diff
@@ -801,9 +815,16 @@ You can define 2 items (1 Switch, 1 Number) mapping to the same channel, see exa
 |          |illumination   |String   |yes      |Current illumination: dark/twilight/bright                           |
 |          |vibration      |Switch   |yes      |ON: Vibration detected                                               |
 |          |charger        |Switch   |yes      |ON: USB charging cable is connected external power supply activated. |
+|          |notionActive   |Switch   |yes      |ON: Motion detection is currently active                             |
```
Comment (fwolter) [inline bundles/org.openhab.binding.shelly/README.md:818]:
```suggestion
|          |motionActive   |Switch   |yes      |ON: Motion detection is currently active                             |
```

File: bundles/org.openhab.binding.shelly/README.md [hunk 3]
```diff
@@ -801,9 +815,16 @@ You can define 2 items (1 Switch, 1 Number) mapping to the same channel, see exa
...
 |          |lastUpdate     |DateTime |yes      |Timestamp of the last update (any sensor value changed)              |
 |battery   |batteryLevel   |Number   |yes      |Battery Level in %                                                   |
 |          |lowBattery     |Switch   |yes      |Low battery alert (< 20%)                                            |
+|device    |sensorSleepTime|Number   |no       |Specifies the number of sec the sensor should not report events      ]
+
+Use case for the 'sensorSleepTime': 
+You have a Motion controlling your light. 
+You switch off the light and want to leave the room, but the motion sensor immediately switches light back on.
+Using 'sensorSleepTime' you could suspress motion events while leaving the room, e.g. for 5sec and the light doesn's switch on.
```
Comment (fwolter) [inline bundles/org.openhab.binding.shelly/README.md:827]:
```suggestion
Using 'sensorSleepTime' you could suppress motion events while leaving the room, e.g. for 5sec and the light doesn't switch on. 
```

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java [hunk 4]
```diff
@@ -317,4 +321,31 @@ public int getRollerFav(int id) {
...
+                return substringBetween(getString(version), "/", "@");
+            }
+
+            if (version.indexOf('/') != version.lastIndexOf('/')) {
+                // includes 2 /, e.g. beta
+                return substringAfterLast(version, "/");
+            }
+        }
+        return "";
+    }
```
Comment (fwolter) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java:341]: It seems more reasonable to extract the version by using regular expressions. That would increase code maintainability.

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/ShellyHandlerFactory.java [hunk 5]
```diff
@@ -141,8 +143,12 @@ public boolean supportsThingType(ThingTypeUID thingTypeUID) {
...
     }
 
-    public Map<String, ShellyBaseHandler> getThingHandlers() {
-        return deviceListeners;
+    public Map<String, ShellyManagerInterface> getThingHandlers() {
+        Map<String, ShellyManagerInterface> handlers = new TreeMap<>();
+        for (Map.Entry<String, ShellyBaseHandler> bh : deviceListeners.entrySet()) {
+            handlers.put(bh.getKey(), bh.getValue());
+        }
+        return handlers;
```
Comment (fwolter) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/ShellyHandlerFactory.java:151]:
```suggestion
    public Map<String, ShellyManagerInterface> getThingHandlers() {
        return new TreeMap<>(deviceListeners);
```

Comment (fwolter) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/ShellyHandlerFactory.java:151]: Is this method used anywhere? What's the reason for using a TreeMap?

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java [hunk 6]
```diff
@@ -242,25 +243,42 @@ private boolean initializeThing() throws ShellyApiException {
...
         tmpPrf.updateFromStatus(tmpPrf.status);
         updateProperties(tmpPrf, tmpPrf.status);
         checkVersion(tmpPrf, tmpPrf.status);
+        if (config.eventsCoIoT && (tmpPrf.settings.coiot != null) && (tmpPrf.settings.coiot.enabled != null)) {
+            String devpeer = getString(tmpPrf.settings.coiot.peer);
+            String ourpeer = config.localIp + ":" + ShellyCoapJSonDTO.COIOT_PORT;
+            if (!tmpPrf.settings.coiot.enabled || (profile.isMotion && devpeer.isEmpty())) {
+                try {
+                    api.setCoIoTPeer(ourpeer);
+                    logger.info("{}: CoIoT peer updated to {}", thingName, ourpeer);
```
Comment (fwolter) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java:272]:
```suggestion
                    logger.debug("{}: CoIoT peer updated to {}", thingName, ourpeer);
```

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java [hunk 7]
```diff
@@ -530,6 +562,25 @@ private void fillDeviceStatus(ShellySettingsStatus status, boolean updated) {
...
+    /**
+     * Check if device has restarted and needs a new Thing initialization
+     *
+     * @return true: restart detected
+     */
+
+    private boolean checkRestarted(ShellySettingsStatus status) {
+        if (profile.isInitialized()
+                && ((status.uptime < stats.lastUptime) || (!profile.status.update.oldVersion.isEmpty()
+                        && !status.update.oldVersion.equals(profile.status.update.oldVersion)))) {
```
Comment (fwolter) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java:574]: Please try to overcome putting more brackets as necessary around expressions.

There's an eclipse save action to remove unncessary brackets:
grafik

```suggestion
                && (status.uptime < stats.lastUptime || !profile.status.update.oldVersion.isEmpty()
                        && !status.update.oldVersion.equals(profile.status.update.oldVersion))) {
```

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/util/ShellyUtils.java [hunk 8]
```diff
@@ -268,6 +269,19 @@ public static DateTimeType getTimestamp(String zone, long timestamp) {
...
+    public static String getTimestamp(DateTimeType dt) {
+        return dt.getZonedDateTime().toString().replace('T', ' ').replace('-', '/');
+    }
+
+    public static String convertTimestamp(long ts) {
+        if (ts == 0) {
+            return "";
+        }
+        String time = DateTimeFormatter.ofPattern(DateTimeType.DATE_PATTERN)
+                .format(ZonedDateTime.ofInstant(Instant.ofEpochSecond(ts), ZoneId.systemDefault()));
```
Comment (fwolter) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/util/ShellyUtils.java:281]: You could make the formatter a constant.

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties [hunk 9]
```diff
@@ -42,6 +42,8 @@ message.roller.favmissing = Roller position favorites are not supported by insta
 # Device
 channel-type.shelly.deviceName.label = Device Name
 channel-type.shelly.deviceName.description = Symbolic Device Name as configured in the Shelly App.
+channel-type.shelly.sensorSleepTime.label = Sensor Sleep Time
+channel-type.shelly.sensorSleepTime.description = The will not send notifications and will not perform actions until the specified time expires. (0=disable)
```
Comment (fwolter) [inline bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties:46]:
```suggestion
channel-type.shelly.sensorSleepTime.description = The sensor will not send notifications and will not perform actions until the specified time expires. (0=disable)
```

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties [hunk 10]
```diff
@@ -68,3 +72,43 @@ channel-type.shelly.ledPowerDisable.description = ON: The power status LED will
...
+message.manager.coiot.new-peer = CoIoT mode/address will be set to {0}.
+message.manager.coiot.mode-mcast = The device starts sending CoIoT updates using IP Multicast.<br/>Please make sure that your network setup supports Multicast routing when devices are on different IP subnets.
+message.manager.coiot.mode-peer = The device will no longer send IP Multicast CoIoT updates to the network, just to the openHAB host.
+
+message.manager.fwupdate.initiated = Firmware update initiated, device returned status {0}
+message.manager.fwupdate.confirm = Do not power-off or restart device while updading the firmware!
+message.manager.fwupdate.info = Update firmware (deviceType={0}, version={1}, url={2})
+message.manager.fwupdate.failed = Firmware updated failed: {0}
+message.manager.fwupdate.notfound = Unable to find firmware for device type {0}, version={1} (url={2})
+message.manager.fwupdate.nofile ) No firmware files found for device type {0}
```
Comment (fwolter) [inline bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties:113]:
```suggestion
message.manager.fwupdate.nofile = No firmware files found for device type {0}
```

**Review by fwolter — CHANGES_REQUESTED** [review]
There are some checkstyle warnings, which could be fixed easily.

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java (hunk 4 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java:341]: I not familiar with regular expressions, would need help

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 6 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java:272]: I wanto to keep this as info, this is a one-time message when the binding changes the peer. Afterwards the device keeps that setting (until the user changes it on the device)

File: bundles/org.openhab.binding.shelly/README.md (hunk 1 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/README.md:808]: remove, this belongs to the next PR

File: bundles/org.openhab.binding.shelly/README.md (hunk 2 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/README.md:818]: fixed

File: bundles/org.openhab.binding.shelly/README.md (hunk 3 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/README.md:827]: fixed

Comment (markus7017) [reply bundles/org.openhab.binding.shelly/README.md]: fixed

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/ShellyHandlerFactory.java (hunk 5 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/ShellyHandlerFactory.java:151]: changed to HashMap<>

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 7 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java:574]: I changed the setting, saved the file, but brackets still stay
?

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/util/ShellyUtils.java (hunk 8 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/util/ShellyUtils.java:281]: changed

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties (hunk 10 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties:113]: fixed

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties (hunk 9 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties:46]: fixed

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/util/ShellyUtils.java (hunk 8 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/util/ShellyUtils.java:281]: done

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 7 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java:574]: option activated, but adjustment was not automatically performed, so I changed this one manually
I made my experiences in the past with missing brackets..., so at some time I decided to take the safe path :-)

Comment (fwolter) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java:574]: You need to make some change in the file to trigger the save actions.

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java (hunk 4 above)
Comment (fwolter) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java:341]: It's not that difficult. You could take a look at other bindings: https://github.com/openhab/openhab-addons/search?q=pattern.compile

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 6 above)
Comment (fwolter) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java:272]: OK

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 7 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java:574]: I did, I'll have an eye on it

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java (hunk 4 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java]: yup, I know how this works in general, but finding the correct pattern...
I found this page: https://regex101.com, realy nice, allows you to test online, but also decodes a regex and shows the logic

Implemented regex search

**Comment by markus7017** [comment]
@fwolter Review changes are applied, was able to fix 2 more bugs, nothing open know at the moment

File: bundles/org.openhab.binding.shelly/README.md (hunk 1 above)
Comment (fwolter) [reply bundles/org.openhab.binding.shelly/README.md:808]: Did you push your change?

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java [hunk 11]
```diff
@@ -317,4 +324,31 @@ public int getRollerFav(int id) {
...
+                // e.g. 20210226-091047/v1.10.0-rc2-89-g623b41ec0-master
+                return matcher.group(0);
+            }
+            if (version.contains("@")) { // standard format
+                return substringBetween(getString(version), "/", "@");
+            }
+            if (version.indexOf('/') != version.lastIndexOf('/')) {
+                // includes 2 /, e.g. beta
+                return substringAfterLast(version, "/");
+            }
```
Comment (fwolter) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java:341]: You might combine these three variants in one regex. Do you have samples of the other two?

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties (hunk 10 above)
Comment (fwolter) [reply bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties:113]: Did you push your change?

**Review by fwolter — CHANGES_REQUESTED** [review]
There are two zip files in this PR.

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java (hunk 11 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java:341]: ok, I removed if () 2+2. In general the pattern should catch them all, but do to the fact I do not have all variants I left them in as fallback.

**Comment by markus7017** [comment]
@fwolter Changes applied a) regex b) not the smartes idea to work on a copy of the documents :-), zip files removed

**Review by fwolter — APPROVED** [review]
LGTM

Great that you figured out the regex!

**Comment by markus7017** [comment]
@fwolter We spoke abozt the regex expression, could you help to define the pattern, which extracts the version, but also -rc2" as sub identifier, samples for version strings:
```
20210226-072307/v1.1.0@f31e1d2b
20201124-092159/v1.9.0@57ac4ad8
20210115-103919/v1.9.4@e2732e05
20210226-115010/v1.9.6-dimmer2@1873d25e
20210122-154345/v1.10.0-rc1@00eeaa9b
20210303-090126/v1.10.0-rc2-153-gd87589bc6-master
20210209-101435/v1.10.0-rc2-g0cf9ee3-heads/v1.10.0-rc2
```

Currently I have the following pattern: "v\\d+\\.\\d+\\.\\d+"
things gets the version, but strips the "-rc1" or "-rc2" or "-dimmer2"

Feedback welcome

**Comment by fwolter** [comment]
`v\d+\.\d+\.\d+(-[a-z0-9]*)?` should work.

**Pull request merged**